### DISCUSSION TITLE
vsgQt refactor to support multiple windows & assocaited updates to VSG and vsgExamples

### DISCUSSION THREAD

**Discussion by robertosfield** [vsg-dev/VulkanSceneGraph#862] [General]
Hi All,

I have been working on vsgQt to enable support for multiple windows including within QMdiArea, to achieve this I have had to refactor the vsgQt backend, and also had to make some updates to the core VSG's vsg::WindowTraits to make this possible, so both the VSG and vsgExamples repos have been updated.

The changes to VSG are to add a Viewer::removeWindow(..) method and replace the vsg::WindowTraits::shareWindow with vsg::WindowTraits:device which are now merged with VSG master, the changes are:

[VulkanSceneGraph changes](https://github.com/vsg-dev/VulkanSceneGraph/pull/861/files)

I also had to update a couple of examples in vsgExamples replacing the WindowTraits::shareWindow usage with WindowTraits::device, these changes may be what you'll need to apply to your own applications if you have been using this feature:

[ vsgExample changes](https://github.com/vsg-dev/vsgExamples/pull/229/files)

The big changes are to vsgQt, in terms of public interface vsgQt::ViewerWindow has been renamed vsgQt::Window, but for now you can still use vsgQt::ViewerWindow naming as I've added a using ViewerWindow = vsgQt::Window.  The public API of vsgQt::Window is mostly unchanged but the back-end has substantially changed.  Finally there are two new examples - vsgqtmdi and vsgqtwindows that illustrate multi-document area with multiple sub windows, and multiple separate windows respectively. The changes are:

[vsgQt changes](https://github.com/vsg-dev/vsgQt/pull/27)

The backend to vsgQt::Window has been completely rewritten, dropping the native Qt Vulkan Surface/Swapchain support entirely and falling back to using the VSG's Vulkan Surface./Swapchain support.  I had to make this change as I could find a way to shared vkDevice between QWindow as Qt has chosen to "simplify" their Vulkan support by tying vkDevice per QWindow. I could find way around this for Qt5 or Qt6, it's a really bone headed design decision from Trolltech as it severally hobbles Qt applications wanting to leverage all of Vulkan capabilities.  

This Qt design decision is not an issue for simple applications with a single Vulkan graphics window, but for multiple Windows it forces one to have multiple logical vkDevice and with it forces users to duplicate data between Windows.  As the VSG is designed to scale properly without such restrictions it was a show stopper so I had no choice but to drop use of Qt's native Vulkan support and use the VSG itself to provide this, with Qt just providing the basic Windowing support.

I have only tested under Linux so far.  The code paths look like they should work under Windows, but I don't think the VSG's macOS support doesn't yet support it so will need to be implemented.  Under Windows we'll need to do some testing and debug if necessary.  I have family time lined up tomorrow through to next Tuesday so will look to get a Windows build next week if no one else gets to it first.

The nice thing about using the VSG backend is that all the WindowTraits settings that can be used with native VSG windowing works now with vsgQt::Window, so enabling MSAA, depth/image formats, Vulkan features etc. all work in the same way as you'll see in the vsgExamples programs demo'ing various features.

I have created two new vsgQt examples, the first [vsgqtwindows](https://github.com/vsg-dev/vsgQt/blob/master/examples/vsgqtwindows/main.cpp) works a bit like vsgwindows example creating multiple windows, each can have their own model or share as we see in this screenshot (note 8 sample MSAA is enabled with the --samples 8):

`vsgqtwindows models/teapot.vsgt --samples 8
`

![Screenshot_20230629_193639](https://github.com/vsg-dev/VulkanSceneGraph/assets/1327952/54f39c03-a7f5-412b-af16-cc0fe61ed189)

The second new example is [vsgqtmdi](https://github.com/vsg-dev/vsgQt/blob/master/examples/vsgqtmdi/main.cpp) with demonstrates use of QMdiArea to manage multiple sub windows:

`$ vsgqtmdi models/openstreetmap.vsgt
`
![Screenshot_20230629_193012](https://github.com/vsg-dev/VulkanSceneGraph/assets/1327952/8ff66d0a-7b64-41e3-aabf-2ce773fbd5e5)

All this code is hot off the presses and needs community testing of VSG, vsgQt and vsgExamples master.  Once it looks like things are working well across platforms I'll tag point releases for each of these projects.  Let me know of success/failures on this thread.

**Comment by WdsPjm** [comment]
hello   To add models from the disk when clicking a button in the interface, it seems that the initializeCallback function is executed every time a model is loaded. Is this the correct approach?

**Reply by robertosfield** [reply]
None of the vsgQt examples have code for dynamically loading models so I can only presume you've add this.  I can't comment comment on behavior of unknown code.

**Comment by WdsPjm** [comment]
I have made modifications to your vsgQtViewer and would like to implement the functionality of dynamically loading models. I would like to confirm whether this framework supports such dynamic loading.

**Reply by robertosfield** [reply]
It's possible to do dynamic loading - there is even a [vsgdynamicloading](https://github.com/vsg-dev/vsgExamples/blob/master/examples/threading/vsgdynamicload/vsgdynamicload.cpp) example :-)

Dynamic loading isn't quite a straight forward as just loading a model and compiling the scene, my plan has to be write an example in vsgQt to illustrate what is required, along the lines of vsgdynamivloading but I'm not a Qt developer so am having to learn things bit by bit with vsgQt.

When implementing dynamic loading there will be no need to call the initWindows/initViewer methods of the vsgQt::Window, these methods should only ever be called during set up of the window.

Perhaps you could share your modified vsgQtViewer code so we can see what you have done, and suggest how it should be changed.

**Reply by WdsPjm** [reply]
Thank you for your response. If I can successfully complete it, I will submit a pull request.

**Reply by robertosfield** [reply]
Create a branch of your vsgQt and create a new example to test out what you want to do, and if it fails then just point us to it and we can make suggestions.  Advising blind on this type of stuff is really impractical.

**Comment by WdsPjm** [comment]
Thank you very much. I have resolved my issue.

**Reply by robertosfield** [reply]
Good to hear, could you share how you have resolved it?

**Comment by WdsPjm** [comment]


**Reply by robertosfield** [reply]
Screenshots? Sorry but that's a really painful way to try and review 3rd party code.  Jumping through  lots of hoops to help you is beyond what I can do with free support.  Please make it easy for us, I have a lot of work on my plate.

This topic has also drifted off the original topic and rather discussing my work on multi-windows, it's now a WdsPjm support thread.  Could you please start a new thread, and if possible come up with human readable user name as a collection of letters is really hard for others to keep track of who's who, a sludge of letters is just confusing for everyone else except for you.

**Comment by WdsPjm** [comment]
"I'm sorry for the trouble I caused you. I apologize sincerely."

**Comment by WdsPjm** [comment]


**Reply by robertosfield** [reply]
Please start a new thread for any further questions.  As I said above your questions aren't on top for this thread, hijacking a thread to ask questions doesn't help anyone, it lowers the signal to noise ratio on the original thread making it unreadable and confuses everyone who might want to follow your questions.
